### PR TITLE
Add `headers` and `retry_after` to `ModelHTTPError`, populate from all provider SDKs

### PR DESCRIPTION
**Pull request by dmontagu**
## Motivation

David Montague (Pydantic / Logfire) raised this in Slack:

> Would it be reasonable to proxy the headers through on `ModelHTTPError`? … I can't pass through the `retry-after` in internal code because we don't forward the response details.

DouweM confirmed: *"Ah yeah agree we should, `Model{API,HTTP}Error` currently only holds body — PR welcome".*

The concrete use case is downstream code that catches a 429 and wants to honour the `Retry-After` header:

```python
try:
    result = agent.run_sync(prompt)
except ModelHTTPError as exc:
    if exc.status_code == 429:
        raise HTTPException(
            status_code=429,
            detail="The AI service is temporarily rate limited. Please try again shortly.",
            headers={"Retry-After": str(exc.headers.get("retry-after", ""))},
        )
    raise
```

### `ModelHTTPError.headers: dict[str, str] | None`

Response headers from the provider SDK, with keys **lowercased** so `exc.headers.get('retry-after')` works predictably regardless of provider casing. `None` when the provider doesn't expose HTTP headers (gRPC-based providers, synthesised errors).

### `ModelHTTPError.retry_after: float | None`

Convenience property that parses the `Retry-After` header, handling both the integer delta-seconds form (`"60"`) and the HTTP-date form (`"Thu, 01 Jan 2099 00:00:00 GMT"`). Returns `None` when the header is absent or unparseable.

## Design decisions

**Why `ModelHTTPError` only and not `ModelAPIError`?**

`ModelAPIError` is the base for non-HTTP failures (connection errors, timeouts, DNS failures) where there is no HTTP response and therefore no response headers. Adding `headers` there would be misleading — callers would need to null-check it even for the common non-HTTP case. `ModelHTTPError` represents an actual HTTP response (4xx/5xx), which always has headers; it's the right scope. (The maintainer's Slack comment referred to both classes in passing, but the reasoning above makes `ModelHTTPError`-only the correct call.)

**Header normalisation to lowercase**

Keys are lowercased on construction (`{k.lower(): v for k, v in headers.items()}`). This makes `exc.headers.get('retry-after')` work even when a provider uses `Retry-After` or `RETRY-AFTER`. httpx already normalises to lowercase, but other SDKs (Cohere's Fern client, botocore) may differ. Lowercasing at the boundary ensures consistent behaviour everywhere.

**Backwards compatibility**

`headers` is keyword-only (`*`) with a default of `None`. All existing positional call sites (`ModelHTTPError(429, 'model-name', body)`) continue to work unchanged. The `__str__` / `message` format is untouched.

**Pickling**

`__reduce__` is updated to return a 3-tuple `(cls, positional_args, state_dict)` and `__setstate__` restores `headers` from `state`. This keeps the positional signature intact while surviving a pickle round-trip with headers.

## Provider coverage

| Provider | Headers source | Notes |
|---|---|---|
| OpenAI | `e.response.headers` (httpx) | ✅ |
| Anthropic | `e.response.headers` (httpx) | ✅ |
| Groq | `e.response.headers` (httpx) | ✅ |
| Mistral | `e.headers` (`MistralError.headers`, httpx) | ✅ |
| Cohere | `e.headers` (`ApiError.headers`, Fern) | ✅ |
| Google | `e.response.headers` (httpx, optional) | ✅ (`None` when `e.response` is `None`) |
| Bedrock | `ResponseMetadata.HTTPHeaders` (botocore) | ✅ |
| HuggingFace | `e.response.headers` (httpx) | ✅ |
| xAI | — | ❌ gRPC transport, no HTTP response |
| OpenRouter | — | ❌ errors synthesised from JSON inside a 200 body |

## Tests

- Existing pickle round-trip tests updated to assert `headers` survives (including `None` and non-`None` cases).
- New unit tests for `ModelHTTPError`: construction, lowercase normalisation, `retry_after` (delta-seconds, HTTP-date, absent, unparseable, no-headers).
- Provider integration tests (unit, no cassettes): OpenAI, Anthropic, Bedrock, xAI (verifies `headers=None` for gRPC).

### Checklist

- [ ] Any **AI generated code** has been reviewed line-by-line by the human PR author, who stands by it.
- [ ] No **breaking changes** in accordance with the [version policy](https://github.com/pydantic/pydantic-ai/blob/main/docs/version-policy.md).
- [ ] **PR title** is fit for the [release changelog](https://github.com/pydantic/pydantic-ai/releases).

[Created via Murmur](https://cloud.murmur.dev/github_app/pydantic/task/github_app%2Fpydantic%2Fw%2Fmurmur-workspace%2Fgithub_oauth%2Fdmontagu%2Fmodelhttperror-headers)

<a href="https://cubic.dev/pr/pydantic/pydantic-ai/pull/6733?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: pydantic_ai_slim/pydantic_ai/exceptions.py [hunk 1]
```diff
@@ -333,14 +336,58 @@ class ModelHTTPError(ModelAPIError):
...
+        Returns ``None`` when the header is absent or cannot be parsed.  The header value
+        is interpreted first as an integer number of seconds, then as an
+        `HTTP-date <https://httpwg.org/specs/rfc9110.html#http.date>`_ string.
+        """
+        if self.headers is None:
+            return None
+        raw = self.headers.get('retry-after')
+        if raw is None:
+            return None
+        try:
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic_ai_slim/pydantic_ai/exceptions.py:380]: Fixed in 81c3b9de. The integer branch now catches `OverflowError` alongside `ValueError`, and a `seconds < 0` guard returns `None` for negative values. Two regression tests added: one for `Retry-After: -1` and one for a 310-digit integer that overflows `float()`.

**Comment by dsfaccini** [comment]
David's AICA here: 🔍 Local review suite (automated, pre-merge)

| Wave | Status | Blocking | Required (do) | Warnings (defer/skip) |
|---|---|---|---|---|
| 1 · code-correctness | ✅ closed | 0 | 1 | 3 |
| 2 · tests/cassettes/docs | ✅ closed | 0 | 0 | 2 |
| 3 · cross-cutting/siblings | ✅ closed | 0 | 0 | 2 |

**Verdict:** 🔧 1 required action — was 🔧 4 required; all 4 resolved by the author in `21253079a` (asctime `retry_after` + import guards + header-landing assertions), leaving one new CI-lint issue that push introduced. 0 blocking; 7/7 changed tests green.
Reviewed at 2026-07-25T13:52:50Z · diff `sha256:d901de78bf2d`

<details>
<summary>🔧 1 required (do)</summary>

_Automated review — flagged by Claude, not yet confirmed by a maintainer._

**1. Unnecessary `# type: ignore` fails CI's pyright lint** · `tests/test_embeddings.py` · `TestGoogle.test_embed_error_with_http_response`
`errors.APIError` declares a settable `.response` attribute (runtime `hasattr` = True, default `None`), so `error_with_response.response = resp  # type: ignore[attr-defined]` needs no suppression. Pyright reports it as an **error** (`reportUnnecessaryTypeIgnoreComment`), which the whole-repo pyright lint job treats as a failure — tests pass but CI lint would go red. Also `# type: ignore` isn't the repo's convention (`# pyright: ignore[code]`). Fix: delete the `# type: ignore[attr-defined]` comment.

</details>

<details>
<summary>✅ 4 required action(s) resolved</summary>

_Automated review — flagged by Claude, not yet confirmed by a maintainer._
_The 4 required actions from the prior round, resolved by the author in `21253079a`._

**1. `retry_after` now accepts the asctime HTTP-date form** · `exceptions.py` · `ModelHTTPError.retry_after` — naive parsed date normalised to UTC; `test_model_http_error_retry_after_http_date_asctime` added.
**2. Inline `import httpx` hoisted** · `tests/models/test_huggingface.py` — moved to module-level imports.
**3. `importorskip('openai')` added** · `tests/test_exceptions.py` · `test_model_http_error_headers_provider_openai` — now gates like its 3 siblings.
**4. Header-landing asserted for the divergent providers** · `tests/models/test_cohere.py`, `tests/models/test_mistral.py`, `tests/test_embeddings.py` — cohere (`e.headers`) + mistral (`dict(e.headers)`) via end-to-end `agent.run_sync`; bedrock-embedding (`metadata.get('HTTPHeaders')`) synthetic; google-embedding positive path new test.

</details>

<details>
<summary>⚠️ 4 informational (defer/skip)</summary>

_Automated review — flagged by Claude, not yet confirmed by a maintainer._

**1. OpenRouter streaming `:1192` leaves `headers=None`** · `models/openrouter.py` · `OpenRouterStreamedResponse._validate_response` — **skip (verified not a gap).** Four reviewers flagged it; traced and refuted: the openai SDK raises a plain `APIError` (no `.response`) for in-band SSE errors, which is why the code reads `e.body`. A real HTTP 429 happens at stream-open, caught by the parent `_map_api_errors` (openai.py) that the PR *did* enrich.

**2. `retry_after` HTTP-date branch reads wall clock** · `exceptions.py` · `ModelHTTPError.retry_after` — **skip.** Researched: exceptions cross the Temporal boundary as an `ApplicationError` (only class-name + message survive), so `exc.retry_after` isn't reachable workflow-side; model requests run in activities; `messages.py` already reads wall-clock via `timestamp=default_factory(_now_utc)` on the always-constructed happy path with no caveat. Not a real determinism issue.

**3. Docs prose "returns … as a `float`"** · `docs/models/overview.md` — **skip.** Contract is `float | None`; adjacent code comment + docstring already carry the `None` case. Optional polish.

**4. Provider tests drive private `_map_api_errors`** · `tests/test_exceptions.py` — **skip.** VCR carve-out satisfied; cohere/mistral header-landing now also covered end-to-end via `agent.run_sync` (author's `21253079a`).

</details>

**Comment by dmontagu** [comment]
**Addressed @dsfaccini's review (commit 21253079):**

1. **asctime naive datetime bug fixed** — added `if retry_time.tzinfo is None: retry_time = retry_time.replace(tzinfo=timezone.utc)` in `retry_after`, with regression test `test_model_http_error_retry_after_http_date_asctime` (far-future 2099 asctime-date, asserts result is a positive float not `None`).

2. **`import httpx` hoisted** to module level in `test_huggingface.py` — httpx is a core dep, no lazy import needed.

3. **`pytest.importorskip`** now first statement in `test_model_http_error_headers_provider_openai`; inline `import openai` removed.

4. **Header value assertions** for all four divergent paths:
   - **cohere**: `ApiError(headers={'retry-after': '60', ...})` in `test_cohere.py::test_model_status_error`, asserts `exc.headers.get('retry-after') == '60'`
   - **mistral**: `httpx.Response(headers={'retry-after': '30', ...})` in `test_mistral.py::test_model_status_error`, same assertion pattern
   - **bedrock-embedding**: `'HTTPHeaders': {'retry-after': '5', ...}` in `test_embeddings.py::test_client_error_with_status_code`, asserts `exc.headers.get('retry-after') == '5'`
   - **google-embedding positive path**: new `test_embed_error_with_http_response` — constructs `APIError` with real `httpx.Response` carrying `retry-after: 10`, asserts header propagation through the `dict(e.response.headers)` branch

**Comment by dmontagu** [comment]
The lone remaining required action (unnecessary `# type: ignore[attr-defined]` on `error_with_response.response = resp`) was already removed in commit `a55ae0ac` before this review landed — pushed at 13:51 UTC. CI is rerunning against that commit now.

File: docs/models/overview.md [hunk 2]
```diff
@@ -192,6 +192,45 @@ attributes showing the queue depth and configured limits. The `name` parameter o
 
 <!-- TODO(Marcelo): We need to create a section in the docs about reliability. -->
 
+## Handling HTTP Errors
+
+When a provider returns a 4xx or 5xx response, Pydantic AI raises a
+[`ModelHTTPError`][pydantic_ai.exceptions.ModelHTTPError].  In addition to the
+[`status_code`][pydantic_ai.exceptions.ModelHTTPError.status_code] and
+[`body`][pydantic_ai.exceptions.ModelHTTPError.body], the exception now exposes the
```
Comment (DouweM) [inline docs/models/overview.md:200]: @dmontagu Docs should never say "it _now_ exposes" -- that implies the user has historical context. on what used to be exposed before. I'd just list what is exposed now, and keep that up to date.

File: docs/models/overview.md [hunk 3]
```diff
@@ -192,6 +192,45 @@ attributes showing the queue depth and configured limits. The `name` parameter o
...
+caller's own HTTP client.  A convenience property
+[`retry_after`][pydantic_ai.exceptions.ModelHTTPError.retry_after] parses that header and
+returns the number of seconds to wait as a `float`, handling both the integer
+delta-seconds and HTTP-date formats:
+
+```python {title="handle_rate_limit.py" test="skip" lint="skip"}
+from pydantic_ai import Agent
+from pydantic_ai.exceptions import ModelHTTPError
+
+agent = Agent('openai:gpt-4o')
```
Comment (DouweM) [inline docs/models/overview.md:216]: A more modern model please :)

File: docs/models/overview.md (hunk 2 above)
Comment (dmontagu) [reply docs/models/overview.md:200]: Addressed in #6756 — reworded to just list what the exception exposes, no "now" framing.

File: docs/models/overview.md (hunk 3 above)
Comment (dmontagu) [reply docs/models/overview.md:216]: Addressed in #6756 — switched the example to `openai:gpt-5.2` to match the rest of the docs.
